File: apple/Sources/DuVayCore/Values.swift
// DuVay — attribute value parsing
//
// Native reimplementation of the shared helpers in src/components/utils.js,
// validated against spec/fixtures/values.json.

import Foundation

public enum DuVayValues {

    /// Parse a comma-separated list or a JSON array into trimmed, non-empty entries.
    public static func valueList(_ value: String?) -> [String] {
        let text = (value ?? "").trimmingCharacters(in: .whitespaces)
        guard !text.isEmpty else { return [] }

        if text.hasPrefix("[") {
            if let data = text.data(using: .utf8),
               let parsed = try? JSONSerialization.jsonObject(with: data),
               let array = parsed as? [Any] {
                return array.map { String(describing: $0).trimmingCharacters(in: .whitespaces) }
                    .filter { !$0.isEmpty }
            }
            // Malformed JSON that still looks like a list is salvaged by hand,
            // matching the web behaviour for hand-authored attribute values.
            guard text.hasSuffix("]") else { return [] }
            return text.dropFirst().dropLast()
                .split(separator: ",", omittingEmptySubsequences: false)
                .map {
                    $0.trimmingCharacters(in: .whitespaces)
                        .trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))
                }
                .filter { !$0.isEmpty }
        }

        return text.split(separator: ",", omittingEmptySubsequences: false)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
    }

    /// The finite numbers in a value list, or `fallback` when there are none.
    public static func numberList(_ value: String?, fallback: [Double] = []) -> [Double] {
        let numbers = valueList(value).compactMap(Double.init).filter { $0.isFinite }
        return numbers.isEmpty ? fallback : numbers
    }

    /// Up to two uppercase initials from a name.
    public static func initials(_ value: String?) -> String {
        (value ?? "")
            .split(whereSeparator: \.isWhitespace)
            .compactMap(\.first)
            .prefix(2)
            .map { String($0).uppercased() }
            .joined()
    }

    /// `YYYY-MM-DD` for a local date.
    public static func isoDate(_ date: Date) -> String { DuVayCalendar.iso(date) }

    /// Parse `YYYY-MM-DD` into local midnight, rejecting overflowed dates.
    public static func parseIsoDate(_ value: String?) -> Date? {
        guard let value, let m = try? /^(\d{4})-(\d{2})-(\d{2})$/.wholeMatch(in: value) else { return nil }
        var parts = DateComponents()
        parts.year = Int(m.1)
        parts.month = Int(m.2)
        parts.day = Int(m.3)
        let cal = DuVayCalendar.calendar
        guard let date = cal.date(from: parts) else { return nil }
        let check = cal.dateComponents([.year, .month, .day], from: date)
        guard check.year == parts.year, check.month == parts.month, check.day == parts.day else { return nil }
        return date
    }

    /// Whether two dates fall on the same calendar day.
    public static func isSameDate(_ a: Date?, _ b: Date?) -> Bool {
        guard let a, let b else { return false }
        return DuVayCalendar.calendar.isDate(a, inSameDayAs: b)
    }

    /// Whether a date sits within an inclusive range. Nil bounds are ignored.
    public static func dateInRange(_ date: Date?, min: Date?, max: Date?) -> Bool {
        guard let date else { return false }
        if let min, date < min { return false }
        if let max, date > max { return false }
        return true
    }

    /// A URL safe to navigate to.
    ///
    /// Trust boundary: only http/https survive, plus relative references. This
    /// is what keeps `javascript:` and `data:` payloads out of href sinks, so
    /// the fixture suite pins it on every platform.
    public static func safeUrl(_ value: String?) -> String {
        let raw = (value ?? "").trimmingCharacters(in: .whitespaces)
        guard !raw.isEmpty else { return "" }
        // Control characters are used to smuggle a scheme past naive parsers.
        if raw.unicodeScalars.contains(where: { $0.value <= 0x1f || $0.value == 0x7f }) { return "" }

        guard let url = URL(string: raw, relativeTo: URL(string: "https://duvay.invalid/")) else { return "" }
        guard let scheme = url.scheme?.lowercased() else { return "" }
        return scheme == "http" || scheme == "https" ? raw : ""
    }
}

File: linux/src/values.rs
//! Attribute value parsing.
//!
//! Native reimplementation of the shared helpers in `src/components/utils.js`,
//! validated against `spec/fixtures/values.json`.

use crate::calendar::{self, DateTime};

/// Parse a comma-separated list or a JSON array into trimmed, non-empty entries.
pub fn value_list(value: Option<&str>) -> Vec<String> {
    let text = value.unwrap_or("").trim();
    if text.is_empty() {
        return Vec::new();
    }

    if text.starts_with('[') {
        // Hand-authored attribute values are often not valid JSON, so a
        // bracketed list is salvaged by splitting rather than rejected.
        if !text.ends_with(']') {
            return Vec::new();
        }
        return text[1..text.len() - 1]
            .split(',')
            .map(|item| item.trim().trim_matches(['"', '\'']).to_string())
            .filter(|item| !item.is_empty())
            .collect();
    }

    text.split(',')
        .map(|item| item.trim().to_string())
        .filter(|item| !item.is_empty())
        .collect()
}

/// The finite numbers in a value list, or `fallback` when there are none.
pub fn number_list(value: Option<&str>, fallback: Vec<f64>) -> Vec<f64> {
    let numbers: Vec<f64> = value_list(value)
        .iter()
        .filter_map(|item| item.parse::<f64>().ok())
        .filter(|n| n.is_finite())
        .collect();
    if numbers.is_empty() { fallback } else { numbers }
}

/// Up to two uppercase initials from a name.
pub fn initials(value: Option<&str>) -> String {
    value
        .unwrap_or("")
        .split_whitespace()
        .filter_map(|part| part.chars().next())
        .take(2)
        .collect::<String>()
        .to_uppercase()
}

/// Parse `YYYY-MM-DD` into midnight, rejecting overflowed dates.
pub fn parse_iso_date(value: Option<&str>) -> Option<DateTime> {
    let text = value?;
    if text.len() != 10 {
        return None;
    }
    let dt = calendar::date(text)?;
    Some(DateTime { hour: 0, minute: 0, second: 0, ..dt })
}

/// Whether two dates fall on the same calendar day.
pub fn is_same_date(a: Option<&DateTime>, b: Option<&DateTime>) -> bool {
    match (a, b) {
        (Some(a), Some(b)) => a.year == b.year && a.month == b.month && a.day == b.day,
        _ => false,
    }
}

/// Whether a date sits within an inclusive range. `None` bounds are ignored.
pub fn date_in_range(date: Option<&DateTime>, min: Option<&DateTime>, max: Option<&DateTime>) -> bool {
    let Some(date) = date else { return false };
    if let Some(min) = min {
        if date < min {
            return false;
        }
    }
    if let Some(max) = max {
        if date > max {
            return false;
        }
    }
    true
}

/// A URL safe to navigate to.
///
/// Trust boundary: only http/https survive, plus relative references. This is
/// what keeps `javascript:` and `data:` payloads out of link sinks, so the
/// fixture suite pins it identically on every platform.
pub fn safe_url(value: Option<&str>) -> String {
    let raw = value.unwrap_or("").trim();
    if raw.is_empty() {
        return String::new();
    }
    // Control characters are used to smuggle a scheme past naive parsers.
    if raw.chars().any(|c| (c as u32) <= 0x1f || (c as u32) == 0x7f) {
        return String::new();
    }

    // A scheme is `ALPHA *( ALPHA / DIGIT / "+" / "-" / "." ) ":"` appearing
    // before any `/`, `?` or `#`. Anything without one is a relative reference
    // and therefore inherits the page's own (http/https) origin.
    let scheme_end = raw.find(':');
    let path_start = raw.find(['/', '?', '#']);
    let has_scheme = match (scheme_end, path_start) {
        (Some(colon), Some(path)) => colon < path,
        (Some(_), None) => true,
        _ => false,
    };
    if !has_scheme {
        return raw.to_string();
    }

    let scheme = raw[..scheme_end.unwrap()].to_ascii_lowercase();
    let valid = !scheme.is_empty()
        && scheme.starts_with(|c: char| c.is_ascii_alphabetic())
        && scheme.chars().all(|c| c.is_ascii_alphanumeric() || matches!(c, '+' | '-' | '.'));
    if valid && (scheme == "http" || scheme == "https") {
        raw.to_string()
    } else {
        String::new()
    }
}
